File: leetcode_java/java/com/goinhn/leetcode/_1_100/_1_10/Sol1.java
package com.goinhn.leetcode._1_100._1_10;

/**
 * 1. 两数之和
 *
 * @author goinhn
 * @date 2020-08-29
 */
public class Sol1 {

    /**
     * 直接使用两遍循环判断是否有和等于目标值的数字
     * <p>
     * time:O(n^2)
     * space:O(1)
     */
    private static class Solution {
        public int[] twoSum(int[] nums, int target) {
            for (int i = 0; i < nums.length; i++) {
                int another = target - nums[i];
                for (int j = i + 1; j < nums.length; j++) {
                    if (nums[j] == another) {
                        return new int[]{i, j};
                    }
                }
            }

            throw new IllegalArgumentException("no two sum numbers");
        }
    }


}

File: leetcode_go/question/_1_100/_1/sol_1.go
// Package _1 两数之和
// @author goinhn
// @date 2022/8/15
package _1

// 暴力法
// time:O(n^2)
// space:O(1)
func twoSum1(nums []int, target int) []int {
	if nums == nil || len(nums) == 0 {
		return []int{}
	}

	for i, v := range nums {
		another := target - v
		for j := i + 1; j < len(nums); j++ {
			if another == nums[j] {
				return []int{i, j}
			}
		}
	}

	return []int{}
}

// hash table
// time:O(n)
// space:O(n)
func twoSum2(nums []int, target int) []int {
	if nums == nil || len(nums) == 0 {
		return []int{}
	}

	hashTable := map[int]int{}
	for i, v := range nums {
		j, ok := hashTable[target-v]
		if ok {
			return []int{i, j}
		}
		hashTable[v] = i
	}

	return []int{}
}
